### DISCUSSION TITLE
Help with enabling ligatures on nvim invoke/disabling on nvim exit

### DISCUSSION THREAD

**Discussion by rivenirvana** [kovidgoyal/kitty#7695] [Q&A] [answered]
I'm trying to make kitty behave such that ligatures are disabled while at the prompt and automatically set to `cursor` strategy while inside nvim. I have confirmed that my font supports the 3 `disable_ligatures` strategies. From what I read from the docs, disabling it at the conf then making a watcher for nvim seemed to be the appropriate approach, but I have a couple of questions. Here's what I got so far:

---

- Relevant kitty.conf settings:
	```
	allow_remote_control	yes
	disable_ligatures		always
	watcher 				ligature.py
	```
- ~/.config/kitty/ligature.py:
	```py
	from typing import Any, Dict

	from kitty.boss import Boss
	from kitty.window import Window

	def on_cmd_startstop(boss: Boss, window: Window, data: Dict[str, Any]) -> None:
    	print(str(data))
		if data.get('cmdline').startswith('nvim'):
			print('command match')
			if data.get('is_start'):
				print('enabling ligatures')
				boss.call_remote_control(window, ('disable_ligatures', f'--match=id:{window.id}', 'cursor'))
			else:
				print('disabling ligatures')
				boss.call_remote_control(window, ('disable_ligatures', f'--match=id:{window.id}', 'always'))
	```
- Changing `'nvim'` to `''` temporarily to hit the code paths, these are the logs when invoking and quitting nvim:
	```
	{'is_start': True, 'time': 1.832512439, 'cmdline': '', 'exit_status': 0}
	command match
	enabling ligatures
	{'is_start': True, 'time': 1.870817722, 'cmdline': '', 'exit_status': 0}
	command match
	enabling ligatures
	[1.952] [PARSE ERROR] Escape codes to resize text area are not supported
	{'is_start': False, 'time': 12.749349619, 'cmdline': '', 'exit_status': 0}
	command match
	disabling ligatures
	```
---

1. Why is `cmdline` an empty string as indicated by `{'is_start': True, 'time': 1.832512439, 'cmdline': '', 'exit_status': 0}`? Is there a better way to detect nvim starting?
2. Based from the logs, it seems like command start is being detected twice. Is that intended, a bug, or a possible misconfig on my setup?

**Comment by rivenirvana** [comment]
I should note that I'm on zsh 5.9 and found out that it works just fine on fish, both questions becoming irrelevant. So leaning towards a misconfig on my end, still trying to figure out where.

**Comment by rivenirvana** [comment]
After a bit more digging around, I found out that:
- Issue 1 seems to happen because of powerlevel10k. Disabling it makes cmdline have a value again.
- Issue 2 seems to be because of kitty's zsh integration. Setting `shell_integration disabled` does not print a duplicate data print at command start/stop. ~~BUT:~~
~~- New issue 3, seems to me that the cmdline values were not being carried over through the intervention of zsh-integration. By removing the duplicates through `shell_integration disabled`, the execution of the watcher file shows that it does not see a cmdline value. In both cases, ligatures do not get enabled inside neovim.~~

Log when invoking kitty and launching/exiting nvim in the spawned window
shell_integration enabled:
```
{'is_start': True, 'time': 1.810739166, 'cmdline': '', 'exit_status': 0}
{'is_start': True, 'time': 1.813670219, 'cmdline': 'nvim', 'exit_status': 0}
[1.933] [PARSE ERROR] Escape codes to resize text area are not supported
{'is_start': False, 'time': 8.585471436, 'cmdline': 'nvim', 'exit_status': 0}
{'is_start': True, 'time': 10.220740602, 'cmdline': '', 'exit_status': 0}
{'is_start': True, 'time': 10.224099459, 'cmdline': 'exit', 'exit_status': 0}
```
---
shell_integration disabled:
```
{'is_start': True, 'time': 1.572482707, 'cmdline': '', 'exit_status': 0}
[1.703] [PARSE ERROR] Escape codes to resize text area are not supported
{'is_start': False, 'time': 3.084502362, 'cmdline': '', 'exit_status': 0}
{'is_start': True, 'time': 3.750409593, 'cmdline': '', 'exit_status': 0}
```

~~Edit: Additionally, invoking `zsh -f` for no rc-files, there is still no cmdline value to be seen. I am confused.~~
Disregard strikethrough lines, that was a misunderstanding on my part due to typo in my watcher file.

**Comment by kovidgoyal** [comment, marked as answer]
Just have nvim set a user variable in kitty on launch and key your watcher to that. See https://sw.kovidgoyal.net/kitty/mapping/#conditional-mappings-depending-on-the-state-of-the-focused-window 

this is doing conditional key mapping but the principle is the same for a watcher.

**Comment by rivenirvana** [comment]
> Just have nvim set a user variable in kitty on launch and key your watcher to that. See [sw.kovidgoyal.net/kitty/mapping#conditional-mappings-depending-on-the-state-of-the-focused-window](https://sw.kovidgoyal.net/kitty/mapping/#conditional-mappings-depending-on-the-state-of-the-focused-window)

That does work very well for this use case, so marking `on_set_user_var()` as answer, thanks! Can even have dynamic behavior through neovim's APIs, as well.

For those that find themselves here with the same issue and want to know the root of it, however, it's just p10k currently not yet supporting kitty's cmdline extension. PR is over here: https://github.com/romkatv/powerlevel10k/pull/2741.
`on_cmd_startstop()` is still a very useful option for interactive/long-running apps that can't set vars like neovim, or that only need kitty logic on start and exit, so hope it gets fixed for other p10k users as well.
